src/main/scala/matwojcik/tagless/tracing/App.scala

```diff
@@ -18,35 +16,50 @@ import matwojcik.tagless.tracing.http4s.Http4sServer
 import org.http4s.server.Server
 
 import scala.concurrent.ExecutionContext
+import scala.io.StdIn
 
 object App extends IOApp with StrictLogging {
 
-  implicit val executionContext: ExecutionContext = ExecutionContext.fromExecutor(Executors.newCachedThreadPool())
   implicit def unsafeLogger[F[_]: Sync]: Logger[F] = Slf4jLogger.unsafeCreate[F]
-  implicit val system: ActorSystem = ActorSystem()
-  implicit val materializer: ActorMaterializer = ActorMaterializer()
 
-  implicit val remoteCaller: RemoteCaller[IO] = RemoteCaller.akkaHttp
+  val ecc = new ExecutionContextAware(ExecutionContext.global)
 
-  val http4sServer: HttpServer[IO, Server[IO]] = Http4sServer.instance(new Http4sServer)
-  val akkaHttpServer: HttpServer[IO, Http.ServerBinding] = AkkaHttpServer.instance(new AkkaHttpServer(Something[IO], RemoteCaller.akkaHttp))
+  override implicit protected val contextShift: ContextShift[IO] = IO.contextShift(ecc)
+  override implicit protected val timer: Timer[IO] = IO.timer(ecc)
+
+  val ecResource = Resource
+    .make(IO(ExecutionContext.fromExecutorService(Executors.newCachedThreadPool())))(ec => IO(ec.shutdown()))
+    .map(new ExecutionContextAware(_))
+
+  implicit val system: Resource[IO, ActorSystem] = Resource.make(IO(ActorSystem()))(sys => IO.fromFuture(IO(sys.terminate())).void)
 
   override def run(args: List[String]): IO[ExitCode] =
-    Program.program(akkaHttpServer)(3)
+    (system, ecResource).tupled.use { case (sys, ec) =>
+      implicit val system = sys
+      implicit val ecc = ec
+      implicit val materializer: ActorMaterializer = ActorMaterializer()
+      implicit val remoteCaller: RemoteCaller[IO] = RemoteCaller.akkaHttp
+
+      val http4sServer: HttpServer[IO, Server[IO]] = Http4sServer.instance(new Http4sServer)
+      val akkaHttpServer: HttpServer[IO, Http.ServerBinding] =
+        AkkaHttpServer.instance(new AkkaHttpServer(Something[IO], RemoteCaller.akkaHttp))
+
+      Program.program(akkaHttpServer)(3).race(IO(StdIn.readLine()))
```

Comment:
`race(stdin)` + ecResource/actor system resource po to żeby dało się wyjść z apki bez ubijania sbt ;P